### PR TITLE
Code coverage support in UI test

### PR DESCRIPTION
**Pull request by TamiTakamiya**
This is for adding code coverage support to our UI tests and enable them in CI. This uses [the new codo coverage support implemented on vscode-extension-tester](https://github.com/redhat-developer/vscode-extension-tester/pull/1234).

Usage:
```bash
yarn install
yarn webpack-dev
yarn coverage-ui-with-mock-lightspeed-server
```
The html report is generated in `out/coverage/ui-test`.  A sample output is shown as follows:
image

Description:

This PR changes the location of lcov.info files. In a CI build, matrix names "test" and "test (macos)" will produce three lcov.info files:

1. unit --> out/coverage/unit/lcov.info
2. ui-test --> out/coverage/ui-test/lcov.info
3. e2e --> out/coverage/e2e/lcov.info

These files are uploaded with codecov-action and merged into one report on the Codecov web UI.

For example the code coverage of playbookGeneration.ts is about 94% when these results are combined:
image

However, if only the e2e test result is selected,
image

The coverage goes down to about 11%:
image

### CODE REVIEW

File: .c8rc [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
+    "reporter": ["text", "html", "lcov"],
```
Comment (TamiTakamiya) [inline .c8rc:2]: Even though vscode-extension-tester has a built-in default c8 options, I needed to prepare this for

- Add "lcov" to "reporter"
- Exclude syntaxHighlighter from coverage calculation.  It is proviided as a bundled js, coverage collection does not work well.

File: Taskfile.yml [hunk 2]
```diff
@@ -146,19 +148,19 @@ tasks:
...
-      - bash -c 'source "${VIRTUAL_ENV}/bin/activate" && yarn run test-ui-current'
-    interactive: true
-  test-ui-old:
-    desc: Run UI tests (oldest vscode)
-    cmds:
-      - task: package
+      - yarn webpack-dev
       - $VIRTUAL_ENV/bin/python3 --version
-      - bash -c 'source "${VIRTUAL_ENV}/bin/activate" && yarn run test-ui-oldest'
+      - ' {{.XVFB}} bash -c ''source "${VIRTUAL_ENV}/bin/activate" && yarn _coverage-ui-with-mock-lightspeed-server'''
```
Comment (TamiTakamiya) [inline Taskfile.yml:155]: Without `{{.XVFB}}`, ui-test (loading Chrome) fails.

File: package.json [hunk 3]
```diff
@@ -955,7 +954,7 @@
     "url": "git+https://github.com/ansible/vscode-ansible.git"
   },
   "scripts": {
-    "compile": "tsc -b && copyfiles -u 4 'node_modules/@ansible/ansible-language-server/out/server/**/*' out",
+    "compile": "tsc -b && tsc -p ./packages/ansible-language-server/tsconfig.json --outDir ./out/server --sourceMap",
```
Comment (TamiTakamiya) [inline package.json:957]: Replaced `copyfiles` with TypeScript compiler + outDir & sourceMap option. By doing this, generated source maps correctly point to actual source location under ./package/ansible-language-server/

File: package.json [hunk 4]
```diff
@@ -966,14 +965,19 @@
...
     "vscode-prepublish": "yarn run webpack",
     "watch": "tsc -b -w",
     "watch-server": "tsc -p ../ansible-language-server --outDir out/server -w",
     "webpack": "yarn run clean && yarn run compile && webpack --mode production --config ./webpack.config.ts",
-    "webpack-dev": "yarn run clean && webpack --mode development --config ./webpack.config.ts",
+    "webpack-dev": "yarn run clean && yarn test-compile && webpack --mode development --config ./webpack.config.ts",
     "webpack:watch": "webpack --mode development --config ./webpack.config.ts --watch",
     "clean": "rimraf out/client out/server out/tsconfig.tsbuildinfo coverage .nyc_output",
-    "test-compile": "yarn run clean && yarn run compile && tsc -p ./",
+    "test-compile": "yarn run clean && yarn run compile && tsc -p ./ --sourceMap",
```
Comment (TamiTakamiya) [inline package.json:980]: Added `--sourceMap` option for getting code coverage.

File: src/features/lightspeed/playbookGeneration.ts [hunk 5]
```diff
@@ -184,7 +184,7 @@ export async function showPlaybookGenerationPage(
         }
 
         const html = await (
-          await require("../utils/syntaxHighlighter")
+          await require(/* webpackIgnore: true */ "../../../../syntaxHighlighter/src/syntaxHighlighter")
```
Comment (TamiTakamiya) [inline src/features/lightspeed/playbookGeneration.ts:187]: `syntaxHighligher` is the code using the [shiki](https://shiki.style/) library, which is used in Playbook Generation feature. It is built with language/color theme definitions using webpack. To workaround an issue with Mocha, it is dynamically imported here. In order to tell webpack not to bundle syntaxHighligher in the client bundle (`extension.js`), a special comment is inserted here.

File: test/index.ts [hunk 6]
```diff
@@ -15,8 +15,8 @@ function setupCoverage() {
     hookRequire: true,
     hookRunInContext: true,
     hookRunInThisContext: true,
-    include: ["out/client/src/**/*.js"],
-    reportDir: "out/coverage",
+    include: ["out/client/src/**/*.js", "out/server/src/**/*.js"],
```
Comment (TamiTakamiya) [inline test/index.ts:18]: For getting ansible-language-server side coverage in e2e test.

File: test/ui-test/lightspeedUiTest.ts [hunk 7]
```diff
@@ -586,6 +586,9 @@ export function lightspeedUIAssetsTest(): void {
     it("Playbook generation command shows an error message when Lightspeed is not enabled", async function () {
       // Open playbook generation webview.
       await workbench.executeCommand("Ansible Lightspeed: Playbook generation");
+      await new Promise((res) => {
```
Comment (TamiTakamiya) [inline test/ui-test/lightspeedUiTest.ts:589]: In serted a 2-sec wait as I saw a test failure occured on macos.

File: webpack.config.ts [hunk 8]
```diff
@@ -3,13 +3,22 @@ import WarningsToErrorsPlugin from "warnings-to-errors-webpack-plugin";
 
 import path from "path";
 
+type EntryType = {
```
Comment (TamiTakamiya) [inline webpack.config.ts:6]: I found c8 code coverage did not work well when js files are bundled.  So for running UI-tests with code coverage, suppress creating bundles of main server/client js files. Still syntaxHighlighter and webview js files need to be built as bundled js files.

**Review by ssbarnea — APPROVED** [review]

**Pull request merged**